### ISSUE TITLE
PMIx_Get provides an incorrect value of `PMIX_LOCAL_PEERS`

### ISSUE THREAD

**Issue by karasevb** [openpmix/openpmix#1551]
Test PMIx program (`test_lpeers`) to get `PMIX_LOCAL_PEERS`:
```C
   ...
    /* get a list of our local peers */
    if (PMIX_SUCCESS != (rc = PMIx_Get(&proc, PMIX_LOCAL_PEERS, NULL, 0, &val))) {
        pmix_output(0, "Client ns %s rank %d: PMIx_Get local peers failed: %s",
                    myproc.nspace, myproc.rank, PMIx_Error_string(rc));
        exit(rc);
    }
    pmix_output(0, "myrank=%d: lpeers=%s\n", myproc.rank, val->data.string);
    ...
```
Log:
```bash
mpirun -np 4 -H node01:2,node02:2 ./test_lpeers
[node01:78436] myrank=1: lpeers=2,3  <--- incorrect
[node01:78435] myrank=0: lpeers=2,3  <--- incorrect
[node02:188182] myrank=2: lpeers=2,3
[node02:188183] myrank=3: lpeers=2,3
```
It is expected that the `lpeers` for procs on `node01` should contain `0,1` ranks, but this is not so.

The problem is in the code block which forms ranks for `PMIX_LOCAL_PEERS` key:
https://github.com/openpmix/openpmix/blob/089187ccb2ff2c10de09c8cc082ec76fadac897e/src/mca/gds/hash/gds_hash.c#L864-L880

The key `PMIX_LOCAL_PEERS` added for the first node will be replaced (removed and added again in lines 875, 880) for the next nodes. Thus, the value of `PMIX_LOCAL_PEERS` always corresponds to the last node, which is incorrect.

**Comment by karasevb** [comment]
That is the possible reason why MLNX CI fails for the multi-server test (`./pmix_test -s 2 -n 2 --job-fence`)

**Comment by karasevb** [comment]
There is patch for this:
```patch
--- a/src/mca/gds/hash/gds_hash.c
+++ b/src/mca/gds/hash/gds_hash.c
@@ -857,28 +857,21 @@ static pmix_status_t store_map(pmix_job_t *trk,
             pmix_list_append(&trk->nodeinfo, &nd->super);
         }
         /* store the proc list as-is */
-        kp2 = PMIX_NEW(pmix_kval_t);
-        if (NULL == kp2) {
-            return PMIX_ERR_NOMEM;
-        }
-        kp2->key = strdup(PMIX_LOCAL_PEERS);
-        kp2->value = (pmix_value_t*)malloc(sizeof(pmix_value_t));
-        if (NULL == kp2->value) {
-            PMIX_RELEASE(kp2);
-            return PMIX_ERR_NOMEM;
-        }
-        kp2->value->type = PMIX_STRING;
-        kp2->value->data.string = strdup(ppn[n]);
-        /* ensure this item only appears once on the list */
-        PMIX_LIST_FOREACH(kp1, &nd->info, pmix_kval_t) {
-            if (PMIX_CHECK_KEY(kp1, kp2->key)) {
-                pmix_list_remove_item(&nd->info, &kp1->super);
-                PMIX_RELEASE(kp1);
-                break;
+        if (0 == strcmp(pmix_globals.hostname, nodes[n])) {
+            kp2 = PMIX_NEW(pmix_kval_t);
+            if (NULL == kp2) {
+                return PMIX_ERR_NOMEM;
             }
+            kp2->key = strdup(PMIX_LOCAL_PEERS);
+            kp2->value = (pmix_value_t*)malloc(sizeof(pmix_value_t));
+            if (NULL == kp2->value) {
+                PMIX_RELEASE(kp2);
+                return PMIX_ERR_NOMEM;
+            }
+            kp2->value->type = PMIX_STRING;
+            kp2->value->data.string = strdup(ppn[n]);
+            pmix_list_append(&nd->info, &kp2->super);
         }
-        pmix_list_append(&nd->info, &kp2->super);
-
         /* save the local leader */
         rank = strtoul(ppn[n], NULL, 10);
         kp2 = PMIX_NEW(pmix_kval_t);
```
The supposed fix for that is checking for equal the current pmix server hostname with `nodes[n]` and store `lpeers` once names are equal, which avoids replacing the key. But this won't work for all cases, because `nodes[n]` may not match the server name in `pmix_globals.hostname`.
The `pmix_globals.hostname` for the server can be obtained in several ways:
- from `gethostname(hostname, PMIX_MAXHOSTNAMELEN-1);` - this is FQDN, which may differ from the short host names provided by the user. Example: Server name "node01.mydomain" and nodes[n] = {"node01", "node02"}.
- from `getenv("PMIX_HOSTNAME")` - the same, may differ from the user's hostlist.
- etc..

@rhc54 @jjhursey
Do you have any thoughts to fix this?

**Comment by rhc54** [comment]
The only way that data can be overwritten is if the same node name is being repeated in the process map. The code directly above what you cite shows why:
```c
        PMIX_LIST_FOREACH(ndptr, &trk->nodeinfo, pmix_nodeinfo_t) {
            if (NULL != ndptr->hostname && 0 == strcmp(ndptr->hostname, nodes[n])) {
                /* we assume that the data is updating the current
                 * values */
                if (NULL == ndptr->hostname) {
                    ndptr->hostname = strdup(nodes[n]);
                }
                nd = ndptr;
                break;
            }
        }
```
We are tracking all the data on a per-node basis. Thus, the "nd" pointer is pointed to a specific node and we are storing/updating the ```PMIX_LOCAL_PEERS``` for that node, not just the local node. You might check to ensure that the process map is correctly created.

Your other point about the hostname is good and one we frequently encounter. For example, in OMPI we have to deal with the case where the allocation comes from the resource manager, but the ```--host``` option is from the user. Once might be an FQDN format while the other is not. Sadly, there is no easy solution. What we do in OMPI is to provide a flag indicating if hostnames are to retain or ignore FQDN. We probably need to add that to PMIx as well.

Once we have that flag, it is the user's responsibility to ensure they tell us the correct setting. If they ask us to ignore FQDN, then we will match "node01.mydomain" with "node01" - otherwise, we will treat those as two different nodes.

**Comment by artpol84** [comment]
@rhc54 

In the code snippet you have provided:
The outer if checks that hostname is non-zero
if (NULL != ndptr->hostname && ...

And later checks if it is zero:
                if (NULL == ndptr->hostname) {
                    ndptr->hostname = strdup(nodes[n]);
                }

Either I’m misreading the code or the inner if statement won’t be ever executed.

**Comment by artpol84** [comment]
@karasevb, in your experiment, which RTE implementation was used?
Was it ORTE/PRRTE or the test server?

**Comment by karasevb** [comment]
@artpol84 ORTE was used. As mentioned above `./test_lpeers` runs thru `mpirun`. OMPI built with the same PMIx version as `test_lpeers`.

**Comment by karasevb** [comment]
As @rhc54 noticed, my guess regarding key rewriting was wrong. But the problem still has. I will investigate why I get the same `lpeers` value for all nodes.

**Comment by rhc54** [comment]
@artpol84 You are quite correct about that "if" clause 😄  Will fix when I get back for SC

**Comment by karasevb** [comment]
The issue should be fixed with commit 2e8d8471efcbf606e3670a2a1b8061b54b8eef95
Testing with the mentioned commit:
```bash
mpirun -np 4 -H node01:2,node02:2 ./test_lpeers
[node01:102092] myrank=2: lpeers=2,3
[node02:89195] myrank=1: lpeers=0,1 <-- correct
[node01:102093] myrank=3: lpeers=2,3
[node02:89194] myrank=0: lpeers=0,1 <-- correct
```

**Comment by jjhursey** [comment]
@karasevb Is there more work to do on this issue or was it all resolved in PR #1552 ?

**Comment by artpol84** [comment]
@karasevb, please confirm

**Comment by karasevb** [comment]
The issue has been resolved by #1552
